tests/platform_test_harness/graphics_platform_test_harness.cpp

```diff
@@ -0,0 +1,676 @@
+            }
+        };
+
+        return std::make_shared<NullPlatformAuthentication>();
+    }
+private:
+    std::thread main_loop_thread;
+    static char const* argv[];
+};
+char const* MinimalServerEnvironment::argv[] = {"graphics_platform_test_harness", nullptr};
+
+std::string describe_probe_result(mg::PlatformPriority priority)
+{
+    if (priority == mg::PlatformPriority::unsupported)
+    {
+        return "UNSUPPORTED";
+    }
+    else if (priority == mg::PlatformPriority::dummy)
+    {
+        return "DUMMY";
+    }
+    else if (priority < mg::PlatformPriority::supported)
+    {
+        return std::string{"SUPPORTED - "} + std::to_string(mg::PlatformPriority::supported - priority);
+    }
+    else if (priority == mg::PlatformPriority::supported)
+    {
+        return "SUPPORTED";
+    }
+    else if (priority < mg::PlatformPriority::best)
+    {
+        return std::string{"SUPPORTED + "} + std::to_string(priority - mg::PlatformPriority::supported);
+    }
+    else if (priority == mg::PlatformPriority::best)
+    {
+        return "BEST";
+    }
+    return std::string{"BEST + "} + std::to_string(priority - mg::PlatformPriority::best);
+}
+
+bool test_probe(mir::SharedLibrary const& dso, MinimalServerEnvironment& config)
+{
+    try
+    {
+        auto probe_fn =
+            dso.load_function<mg::PlatformProbe>("probe_graphics_platform", MIR_SERVER_GRAPHICS_PLATFORM_VERSION);
+
+        auto const result = probe_fn(
+            config.console_services(),
+            *std::dynamic_pointer_cast<mir::options::ProgramOption>(config.options()));
+
+        std::cout << "Probe result: " << describe_probe_result(result) << "(" << result << ")" << std::endl;
+
+        return result > mg::PlatformPriority::dummy;
+    }
+    catch (std::exception const& err)
+    {
+        std::cout << "Probing failed: " << boost::diagnostic_information(err) << std::endl;
+        return false;
+    }
+    catch (...)
+    {
+        std::cout << "Probing failed with broken exception!" << std::endl;
+        return false;
+    }
+}
+
+auto test_display_platform_construction(mir::SharedLibrary const& dso, MinimalServerEnvironment& env)
+-> std::shared_ptr<mir::graphics::DisplayPlatform>
+{
+    try
+    {
+        auto create_display_platform = dso.load_function<mg::CreateDisplayPlatform>(
+            "create_display_platform", MIR_SERVER_GRAPHICS_PLATFORM_VERSION);
+
+        auto display = create_display_platform(
+            env.options(),
+            env.emergency_cleanup_registry(),
+            env.console_services(),
+            env.display_report(),
+            env.logger());
+
+        std::cout << "Successfully constructed DisplayPlatform" << std::endl;
+
+        return display;
+    }
+    catch (std::exception const& err)
+    {
+        std::cout << "Display construction failed: " << boost::diagnostic_information(err) << std::endl;
+        throw;
```

Comment:
Exceptions propagating out of this function are never caught. Although `return {};` would be handled as an error, we should probably also have a top-level catch handler.

Reply:
Now with top-level exception handler.